### ISSUE TITLE
It doesn't support on python3.13

### ISSUE THREAD

**Issue by toufa12** [tensorflow/tensorflow#78774]
### Issue type

Build/Install

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

tf 2.17

### Custom code

Yes

### OS platform and distribution

macos sequoia arm

### Python version

3.13

### Current behavior?

It doesn't work when I want to install via terminal with the installation code. 

### Standalone code to reproduce the issue

```shell
ERROR: Could not find a version that satisfies the requirement tensorflow (from versions: none)
ERROR: No matching distribution found for tensorflow
```

**Comment by mihaimaruseac** [comment]
TensorFlow always supports the new Python version in the release that follows nearly a quarter after the Python release. This is because the TensorFlow release cycle starts before the Python release occurs and cannot be changed mid-flight. If you look at past issues, you see that this has been an issue since at least python 3.8.

It has been suggested to test release candidates, etc., but no improvements have been made on this front.

**Comment by BaseMax** [comment]
Same issue

I am going to uninstall Python 3.13 and install an older python version.

**Comment by muayyad-alsadi** [comment]
python 3.13 is important because it's the default and only version in the latest fedora 41 (released 2024-10-29)
fedora is a major distro and many users are upgrading to the latest version.

I don't expect tf to support every distro. but python 3.13 is the latest released stable version from python foundation and is used by a major distro. it would be a shame not to support it. I believe we need to adjust or sync the process.

those released are scheduled and predictable

**Comment by iseeface** [comment]
still no fix?

**Comment by mihaimaruseac** [comment]
> TensorFlow always supports the new Python version in the release that follows nearly a quarter after the Python release. This is because the TensorFlow release cycle starts before the Python release occurs and cannot be changed mid-flight. If you look at past issues, you see that this has been an issue since at least python 3.8.
> 
> It has been suggested to test release candidates, etc., but no improvements have been made on this front.

This is still true. Until the next release next year, there won't be a change.

**Comment by yuchen001** [comment]
In addition to Python 3.13 not being supported, could someone confirm if Python 3.12 is currently supported?

**Comment by mihaimaruseac** [comment]
It should be (3.12)

**Comment by thebaptiste** [comment]
According to the available wheels on PyPI, it seems rather to be 3.9-3.12 since release 2.16.1
May be the install doc has not been updated...

**Comment by muayyad-alsadi** [comment]
we just need to do the following
when a new stable python is released by PSF just get last tag and rebuild it and push the wheel for that python version.
there is nothing preventing us from doing this simple automation and I'm not aware of any compatibility issue preventing us from doing this.

**Comment by mihaimaruseac** [comment]
Unfortunately the build system for TF is too complicated for this.

There needs to be a lot of work done to make that possible. Plus, TF has a large number of dependencies that also need to start supporting the new version of Python. And, C++ <-> Python interop has a lot of issues with any new version of Python.

**Comment by muayyad-alsadi** [comment]
would you please point to the CI/CD failed build logs showing the claimed incompatibility? I'm not aware of any. This is not compatibility issue but rather process issue.

Also related:

https://github.com/tensorflow/tensorflow/issues/79349

**Comment by svenstaro** [comment]
Here are my build logs: [tensorflow-python313.txt](https://github.com/user-attachments/files/18223075/tensorflow-python313.txt)

**Comment by muayyad-alsadi** [comment]
```
Error in fail: 
Could not find requirements_lock.txt file matching specified Python version.
Specified python version: 3.13
Python versions with available requirement_lock.txt files: 3.9, 3.10, 3.11, 3.12
```

you must be kidding, let's have mutual respect, I can read and this is not the actual build log this is just an early exist resulted from a pinned version that can patched by a single-line change. basically it says "we don't want you to to try to build it on 3.13".

the current status of tensorflow is unacceptable as it's uninstallable on windows, macos and latest fedora linux all with latest stable python from python foundation.

**Comment by svenstaro** [comment]
So, two things:
- This issue is about tensorflow not supporting 3.13. If you read the error message, it quite clearly doesn't support 3.13.
- If you try to actually patch the files (as we tried in Arch), you'll see various failures in parts of tensorflow. I can get you some logs but you can also try it yourself.

**Comment by mihaimaruseac** [comment]
For the last release of TF (2.18), bramnch cut ocurred slightly before Sept 25 (given https://github.com/tensorflow/tensorflow/commit/b4b3e2da6b64a996f418810524a3274a6bef7fa5 to update version number to RC0).

But on Sept 25, `grpcio` was at 1.66.1, which has [no Python 3.13 support](https://pypi.org/project/grpcio/1.66.1/#files).

This is just the first dependency in the chain that needed to be updated before TF is able to support a new version of Python. I didn't check [any other of the direct dependencies from deps.dev](https://deps.dev/pypi/tensorflow/2.18.0/dependencies) (note that the dependency support versions listed there are not the ones available when the branch was cut, not the ones for RC0).

Because the release process for TF is expensive, once a branch is cut, it can only receive minor cherry-picks, not features from master branch.

That being said, `tf-nightly` should have support for a new version of Python much earlier than it does now, as soon as all dependencies support the new version of Python, TF should be able to do so too. But it is now a requirement on the planning for the OSS team, I am no longer TL there and won't make planning statements.

**Comment by jyo64** [comment]
Tensorflow support for Python 3.13 is a deal breaker now. Even where project that have tensorflow as a dependency 😢.

**Comment by jyo64** [comment]
> For the last release of TF (2.18), bramnch cut ocurred slightly before Sept 25 (given [b4b3e2d](https://github.com/tensorflow/tensorflow/commit/b4b3e2da6b64a996f418810524a3274a6bef7fa5) to update version number to RC0).
> 
> But on Sept 25, `grpcio` was at 1.66.1, which has [no Python 3.13 support](https://pypi.org/project/grpcio/1.66.1/#files).
> 
> This is just the first dependency in the chain that needed to be updated before TF is able to support a new version of Python. I didn't check [any other of the direct dependencies from deps.dev](https://deps.dev/pypi/tensorflow/2.18.0/dependencies) (note that the dependency support versions listed there are not the ones available when the branch was cut, not the ones for RC0).
> 
> Because the release process for TF is expensive, once a branch is cut, it can only receive minor cherry-picks, not features from master branch.
> 
> That being said, `tf-nightly` should have support for a new version of Python much earlier than it does now, as soon as all dependencies support the new version of Python, TF should be able to do so too. But it is now a requirement on the planning for the OSS team, I am no longer TL there and won't make planning statements.



I just tried installing all the dependencies manually as mentioned here. The only thing that has failed is tensorflow-io-gcs-filesystem. I think grpcio (the latest version) now supports Python 3.13.

**Comment by muayyad-alsadi** [comment]
>  tensorflow-io-gcs-filesystem. I think grpcio (the latest version) now supports Python 3.13.

you are correct

![image](https://github.com/user-attachments/assets/9ae6face-7596-49e5-9f6e-a87c98e584b8)

which further indicate it's a political decision within tensorflow team (tensorflow-io-gcs-filesystem) not a technical one.
it's a matter of flipping a switch.

>  I'm not aware of any compatibility issue preventing us from doing this. the current state seems careless, lazy and incompetent.

**Comment by mihaimaruseac** [comment]
grpcio support for 3.13 was added after TF branch cut.

I already explained several times that it is expensive to change the build process once the release train has started.

**Comment by mihaimaruseac** [comment]
Google has its own requirements too, it's not "you can just schedule".

You can also send PRs, the build process is 90% in the repo itself.

**Comment by mihaimaruseac** [comment]
There are actually Python 3.13 features that require more work from TF:

- the new free-threaded runtime: https://docs.python.org/3/whatsnew/3.13.html#free-threaded-cpython. Since TF cannot know if it is called from the new runtime or not, we need to ensure that both are supported. But this entails making sure that there are no incompatibilities with the way TF kernels have been written, there might be assumptions that are broken. Hence, this needs a longer baking in period than "just" releasing.
- the new JIT compiler: https://docs.python.org/3/whatsnew/3.13.html#free-threaded-cpython. TF already does a lot of JIT via XLA.
- there are `ctypes` changes that we need to look into: https://docs.python.org/3/whatsnew/3.13.html#free-threaded-cpython

This is on a quick glance on the release notes. I might have missed some other changes and maybe not all of the above are true, TF is too big and I barely scratched the entire surface of it.

**Comment by mihaimaruseac** [comment]
> the problem that next train is scheduled to be after 6 months

That's not true. Releases are quarterly, not biannually.

**Comment by muayyad-alsadi** [comment]
All of those are runtime not build time. Supporting some runtimes is better
than supporting zero runtimes.

**Comment by mihaimaruseac** [comment]
> Supporting some runtimes is better

Yes, and TF supports Python 3.12, 3.11 and so on. More runtimes by February.

**Comment by muayyad-alsadi** [comment]
I was referring to "...the new free-threaded runtime... the new JIT
compiler"

We did not push wheels for py 3.13 because it can be called in an
unsupported rt. Well! If i call it in the new rt it would be my problem.
I'll be sure to call it in one of the supported ways.

The current state in case you don't know there is no usable wheels for
Windows, mac or fedora linux, arch linux ..etc.

How to reproduce? get the latest stable os from any of those vendors,
install the latest stable python. Try to install TF.

**Comment by mihaimaruseac** [comment]
NixOS 24.11 is on Python 3.12. Fedora 41 released with Python 3.13 by default, but it does not block you from installing a previous version. `hatch`, `uv`, `venv` are all tools you could use to manage python installations.

**Comment by mihaimaruseac** [comment]
To be clear, I am still of the opinion that `tf-nightly` should already support 3.13, it's not really ok to delay adding support to it for so long.

**Comment by nbecker** [comment]
> python 3.13 is important because it's the default and only version in the latest fedora 41 (released 2024-10-29) fedora is a major distro and many users are upgrading to the latest version.
> 
> I don't expect tf to support every distro. but python 3.13 is the latest released stable version from python foundation and is used by a major distro. it would be a shame not to support it. I believe we need to adjust or sync the process.
> 
> those releases are scheduled and predictable

There is python 3.12 on Fedora 41, 
sudo dnf install python3.12

**Comment by muayyad-alsadi** [comment]
This reminds me of nvidia/ubuntu mentality. Canonical had one kernel
developer and his only job is to make sure they ship a kernel that works
with nvidia drivers, if not they ship the older kernel.

As long as there are no breaking changes done by PSF in their latest stable
python, we should build and publish the wheels. We should not hold
innovation waiting for church approval.

**Comment by ilcesko** [comment]
I read the whole discussion but I couldn't find an actual ETA for 3.13 support, do we have one ?
thanks

**Comment by mihaimaruseac** [comment]
Next release of TF. Branch cut should occur by mid of February at most and final release at most a month after, assuming no delays.

**Comment by muayyad-alsadi** [comment]
it seems that church priests keep down voting me

![image](https://github.com/user-attachments/assets/74bfe9af-fa7f-40c5-b611-2ee97f8e64a3)

and ignore the community

![image](https://github.com/user-attachments/assets/e9af7f2d-29ed-4fa2-81a6-4f9b1e7ae4b8)

> Next release of TF. Branch cut should occur by mid of February 
 
This is not the first time this happens, we had such tickets before and we will continue to have them when PSF release python 3.14

supporting python 3.13 is not the root cause of the problem.
there should be a way within TF church to request a build of the current stable of version of TF (no new features) against the current stable version of python without version pinning.

> That being said, tf-nightly should have support for a new version of Python much earlier than it does now,

yes, should have, but did not.

**Comment by mihaimaruseac** [comment]
Apologies for the wrong signal sent by :-1:. I wanted a quick feedback to signal that you are repeating the same information, in a way that could be construed as aggressive and against [the code of conduct](https://github.com/tensorflow/tensorflow/blob/master/CODE_OF_CONDUCT.md).

This is not Reddit or similar, so a :-1: has no effect (I got a few myself too, even here). But, taking into account your feedback, I have removed it.

Regarding the rest of your message, I can only reply by repeating what I have already said. In the spirit of making progress here, I will just summarize:

- Stable release of new Python version is impossible to add after branch cut
- Testing RCs of new Python version just before release train starts risks delaying releases for everyone and then can lead to issues down the line (e.g., Python 3.11 docker images for some TF versions are on an RC version even now because of this)
- Release calendar has to take into account multiple priorities and coding freezes. It cannot be shifted around to match the Python release schedule.
- Nightly support of new Python version can be added, but in general it gets delayed. It is the responsibility of the OSS team to do so and they have to juggle quite a large number of priorities. Please assume the best intentions, there is no one that wants the community to not be supported
- I have left the TF OSS team in 2022 and have been only helping here and there to smooth the transition. I do not represent the TF team and I don't know why you are picking on me. I'm just responding to the issue rather than let it stale.

Please make sure to read the code of conduct.

**Comment by muayyad-alsadi** [comment]
Code of conduct? Church here is not a religious reference!! It's a
hierarchical management reference used by Eric S. Raymond in his book "The
Cathedral and the Bazaar" (abbreviated CatB) which is one of the
foundations of OpenSource community (he is the one behind OSI definition).

How come you did not recognize the reference to CatB and OSI? Never heard
of it? Am I that old?

https://en.m.wikipedia.org/wiki/The_Cathedral_and_the_Bazaar

**Comment by medaminezghal** [comment]
I’m using Arch Linux and I can’t update keras to the latest version due to its dependency of tensorflow that is still unusable because it isn’t supported in python 3.13.
So I need to wait until the team make it compatible with python 3.13 to use keras updated.
That’s every year problem. 😑

**Comment by mihaimaruseac** [comment]
Or you can install python3.12 on Arch, you can have multiple versions in parallel.

**Comment by xatier** [comment]
@medaminezghal, you can simply `$ yay python312` to install Python 3.12 on Arch to the path `/usr/bin/python3.12`, just FYI.

Ref: https://wiki.archlinux.org/title/Python#Other_versions

**Comment by mihaimaruseac** [comment]
Also there are tools (like `hatch`) that manage Python installation for you, so "I cannot use it on $platform because it comes with Python 3.13 by default" arguments don't stand.

**Comment by ieshaan12** [comment]
Are there any release candidates for the same? Potentially, some release candidates, I think it breaks a lot of builds where the default python is 3.13. I'm fine downgrading, but its a bit of a hassle nonetheless.

**Comment by mihaimaruseac** [comment]
It doesn't seem to be in nightly yet. And there is no new branch already cut to signal starting a new release (which would be a prerequisite for RCs).

**Comment by VedeshP** [comment]
so is there any other way to install tensorflow as i am using a virtual environment and i have other dependencies already installed - some code already written - if i have to revert and make a new env with older python version it will be a tedious task

**Comment by xatier** [comment]
@VedeshP why don't you dump the venv with `pip freeze` to a `requirementx.txt` file? You may also use `pipreqs` if you want leaf packages only.

**Comment by mihaimaruseac** [comment]
It's really not tedious at all, `pip freeze > requirements.txt` and then in the new environment `pip install -r requirements.txt`

**Comment by VedeshP** [comment]
@xatier @mihaimaruseac 
That's a good idea, never thought about this 
Thank you 
There's one more question 
Shall I use the jupyter kernel made from the global python installed in my computer?

**Comment by xatier** [comment]
@VedeshP please do not go off-topic. You should do your homework on Stack Overflow or other forums instead.

**Comment by medaminezghal** [comment]
The Arch Linux developers has successfully provided a patch to make TensorFlow work with python 3.13

**Comment by xatier** [comment]
@sandeepb20, here you go:

https://gitlab.archlinux.org/archlinux/packaging/packages/tensorflow/-/blob/main/tensorflow-2.18.0-python-3.13-API.patch?ref_type=heads

You can refer to the changes to this repo.
https://gitlab.archlinux.org/archlinux/packaging/packages/tensorflow/-/commits/main?ref_type=HEADS

**Comment by mihaimaruseac** [comment]
> [@medaminezghal](https://github.com/medaminezghal) Can you please provide reference for the patch to make TensorFlow work with python 3.13? And how can we use that? Also when can we expect that tensorflow will officially support python3.13?

I used to say that this would come with the next release, but it seems branch has been cut without nightly supporting Python 3.13.

Most likely the team had other priorities, or forgot to bump support.

**Comment by Kush3007** [comment]
It's been 5 months since the release of 3.13. I doesn't seem like you guys know how long a quarter is.

**Comment by medaminezghal** [comment]
> It's been 5 months since the release of 3.13. I doesn't seem like you guys know how long a quarter is.

Maybe it will be compatible after the release of python 3.15 😅

**Comment by mihaimaruseac** [comment]
CC @learning-to-play since this has been missed from the new release and it's been really active in the past quarter

**Comment by sandeepb20** [comment]
Hi @mihaimaruseac, will 3.13 support be available next quarter? 
Could you provide a tentative timeline? 
Is there any chance we might see support by mid-quarter?

**Comment by mihaimaruseac** [comment]
@learning-to-play can provide timelines, it's his team that does the work

**Comment by sbijarnia-eightfold** [comment]
@learning-to-play Please provide updates. We are blocked on this for a project, quick fix would be highly appreciated 🙌

**Comment by sandeepb20** [comment]
@learning-to-play  Please provide updates. We are blocked on this for a project, quick fix would be highly appreciated 🙌

**Comment by learning-to-play** [comment]
TF 2.19 will not have Python 3.13 support. I informed our PM about this request.

**Comment by crazyguy1908** [comment]
so do we wait till several months after 3.14 release to use tf on the latest version of python??

**Comment by itsXactlY** [comment]
Well, was sooo close to blark about the same, but thanks who ever posted that the arch guys made an build already. <3

**Comment by medaminezghal** [comment]
So tf-nightly 2.20 released without support for python 3.13 and few months later python 3.14 will be released.

**Comment by swiftclouddbs** [comment]
Found out today 3.13 cannot run TF. Numpy not in 3.13 as of a few weeks ago. venv anyone?

**Comment by muayyad-alsadi** [comment]
> So tf-nightly 2.20 released without support for python 3.13 and few months later python 3.14 will be released.

yes, TF team do this every single release and they tell people that PSF made breaking changes in the new python,
when in reality it's just an incompetent in-house build system problem resulted from pinning the version.

when the RC is released they will close this ticket and pretend it's fixed. then we will open a new one for python 3.14 then we wait half a year, then close it again.

**Comment by mihaimaruseac** [comment]
> > So tf-nightly 2.20 released without support for python 3.13 and few months later python 3.14 will be released.
> 
> yes, TF team do this every single release and they tell people that PSF made breaking changes in the new python, when in reality it's just an incompetent in-house build system problem resulted from pinning the version.
> 
> when the RC is released they will close this ticket and pretend it's fixed. then we will open a new one for python 3.14 then we wait half a year, then close it again.

@learning-to-play

**Comment by loqs** [comment]
> TF team do this every single release and they tell people that PSF made breaking changes in the new python,
when in reality it's just an incompetent in-house build system problem.

@muayyad-alsadi can you build against 3.13 without making changes for:
 - _PyArg_NoKeywords related upstream commit https://github.com/python/cpython/commit/be5e8a010341c4d2d28ef53a1baed402ee06466e.
 - _PyObject_VisitManagedDict/_PyObject_ClearManagedDict related upstream commit https://github.com/python/cpython/commit/fc2cb86d210555d509debaeefd370d5331cd9d93.

Or if you already have a working 3.13 build with those changes made you could submit them.

**Comment by medaminezghal** [comment]
@loqs Thanks for providing the latest version 2.19 compatible with Python 3.13 in Arch Linux.

**Comment by jarif87** [comment]
same problem here this is not good

**Comment by Aeonitis** [comment]
I feel like there should be some simple compatibility test which is only ran once per operating system after each version change for such a popular and important repo.

- install/import tensorflow 
- run tensorflow operation 
- assert expected result 

If successful, you know
If unsuccessful, you know

Either way, update docs & confirm the last supported OS version.

Saves a lot of time for the rest of the world using or trying to use it.

**Comment by loqs** [comment]
> I feel like there should be some simple compatibility test which is only ran once per operating system after each version change for such a popular and important repo.

The build is guaranteed to fail for a new major python release due to a missing requirements file for that major version.

**Comment by Lyes-Im** [comment]
Pour les gens qui utilisent Jupyter Notebook sur anaconda, vous pouvez vérifier dans anaconda prompt la version du python utilisée, si vous trouverez V12, vous pouvez facilement travailler avec TensorFlow

**Comment by tyoc213** [comment]
```
uv run --with jupyter jupyter lab
  × No solution found when resolving dependencies:
  ╰─▶ Because all versions of tensorflow have no wheels with a matching Python version tag (e.g., `cp313`) and your project depends on tensorflow, we can conclude that your project's requirements are
      unsatisfiable.

      hint: Pre-releases are available for `tensorflow` in the requested range (e.g., 2.19.0rc0), but pre-releases weren't enabled (try: `--prerelease=allow`)

      hint: Wheels are available for `tensorflow` (v2.19.0) with the following Python ABI tags: `cp39`, `cp310`, `cp311`, `cp312`
```

osx

**Comment by mihaimaruseac** [comment]
> ```
> uv run --with jupyter jupyter lab
>   × No solution found when resolving dependencies:
>   ╰─▶ Because all versions of tensorflow have no wheels with a matching Python version tag (e.g., `cp313`) and your project depends on tensorflow, we can conclude that your project's requirements are
>       unsatisfiable.
> 
>       hint: Pre-releases are available for `tensorflow` in the requested range (e.g., 2.19.0rc0), but pre-releases weren't enabled (try: `--prerelease=allow`)
> 
>       hint: Wheels are available for `tensorflow` (v2.19.0) with the following Python ABI tags: `cp39`, `cp310`, `cp311`, `cp312`
> ```
> 
> osx

As the error message says, Py3.9 to Py3.12 are the only supported for 2.19.

**Comment by loqs** [comment]
Anyone having issues building with python 3.13 from git now https://github.com/tensorflow/tensorflow/commit/c0093bfb21f0e1157e1892cc8273e10831077174 has been merged?

**Comment by mihaimaruseac** [comment]
https://pypi.org/project/tf-nightly/2.20.0.dev20250413/#files has wheels for Python 3.13

**Comment by medaminezghal** [comment]
6 months and the python 3.14 will be released and tensorflow support python 3.13 only in nightly

**Comment by khteh** [comment]
Any timeline for this to be fixed / released?

**Comment by Aeonitis** [comment]
We need to be patient with open source projects.

That said they also need to label which issues they consider "high priority" or "fundamentally important" because it also communicates the weight of their current tasks.

I understand it can be frustrating on both sides.

There seems to be a label called 'prime' but it's applied to none of the issues, maybe start with that?

**Comment by mihaimaruseac** [comment]
> Any timeline for this to be fixed / released?

Next TF release

**Comment by khteh** [comment]
> > Any timeline for this to be fixed / released?
> 
> Next TF release

When? How's the release cycle of this project/repository like?

**Comment by gsandrof66** [comment]
Is there any possible release date? I am thinking to use Torch instead.

**Comment by medaminezghal** [comment]
> Is there any possible release date? I am thinking to use Torch instead.

Approximately, each version released between 3 and 5 months.

**Comment by muayyad-alsadi** [comment]
> Approximately, each version released between 3 and 5 months.

python 3.13 was released by PSF in oct 2024
so it's not a matter of 3-5 months

TF team released TF versions after that and they did not include support for py3.13
 
1. there is no ABI breaking changes in py3.13
2. archlinux team was able to build TF for py3.13 (so it's not a matter of compatibility issues)

it's just TF team pinning versions for internal google processes and pipelines disregarding opensource community.

**Comment by thebaptiste** [comment]
If there were a competition for the worst management by a project of technical changes outside the project (such as Python 3.13 or numpy 2.2), tensorflow would probably be a good candidate for the gold medal.
Of course it's easier to do for a small project but it's also a question of managing priorities and taking into account community needs.

**Comment by loqs** [comment]
>     1. there is no ABI breaking changes in py3.13
> 
>     2. archlinux team was able to build TF for py3.13 (so it's not a matter of compatibility issues)
> 
> 
> it's just TF team pinning versions for internal google processes and pipelines disregarding opensource community.

That is the same claim you made in https://github.com/tensorflow/tensorflow/issues/78774#issuecomment-2713021735 to which I listed the ABI breaking changes that were breaking the build in https://github.com/tensorflow/tensorflow/issues/78774#issuecomment-2715410042. The same ABI compatibility issues that were addressed in https://github.com/tensorflow/tensorflow/commit/c0093bfb21f0e1157e1892cc8273e10831077174. Points one and two are internally inconsistent if there were no ABI breaking changes why would have to apply [tensorflow-2.18.0-python-3.13-API.patch](https://gitlab.archlinux.org/archlinux/packaging/packages/tensorflow/-/blob/main/tensorflow-2.18.0-python-3.13-API.patch)?

**Comment by h-vetinari** [comment]
As an open source maintainer[^1], I've got to say that the entitlement in this thread is just off the charts.

Whether funded by $BIGCORP or run by volunteers in their spare time, you don't get to dictate how a given FOSS project prioritizes its work. You're not paying for anything, and so you're (still) getting what you paid for. Of course, you could rearchitect your own project to switch to another framework or language, but that sure sounds like a lot of work - definitely easier to just demand stuff...

Speaking of, if you think you need python 3.13 today (or freethreading tomorrow, etc.): No library _requires_ python 3.13[^3], and >99.9% of applications could be done equivalently on 3.12 (aside from some minor performance gains[^2]). Even the semi-official 3.13 feature overview [notes](https://realpython.com/python313-new-features/#so-should-you-upgrade-to-python-313):

> So, having seen all these new improvements, should you upgrade to Python 3.13? Like for every release, the answer is a clear and loud **it depends**!

Engineers are used to making trade-offs (c.f. "it depends"), and to collaborate professionally. If you know how to do neither one nor the other, you're probably... not as good of a programmer as you think you are.

[^1]: of a large, cross-platform, rolling [distribution](https://conda-forge.org/), including for our tensorflow [recipe](https://github.com/conda-forge/tensorflow-feedstock/)

[^2]: and if those performance gains have a measurable financial impact, why are you not offering to pay (a portion of that) to your FOSS dependencies to do the work necessary to support a new python version?

[^3]: because it would be project suicide to artificially restrict your reach to ~2.5% of current python users (as measured through overall PyPI downloads; see "Consumer statistics" [here](https://mayeut.github.io/manylinux-timeline) - that's linux-only but has a nice dashboard; for everything else there's [pypinfo](https://github.com/ofek/pypinfo))

**Comment by dsmaynard** [comment]
> If there were a competition for the worst management by a project of technical changes outside the project (such as Python 3.13 or numpy 2.2), tensorflow would probably be a good candidate for the gold medal. Of course it's easier to do for a small project but it's also a question of managing priorities and taking into account community needs.

You're always welcome to contribute to an open-source project yourself rather than complain that others aren't doing what you want (for free).

**Comment by medaminezghal** [comment]
> You're always welcome to contribute to an open-source project yourself rather than complain that others aren't doing what you want (for free).



> Whether funded by $BIGCORP or run by volunteers in their spare time, you don't get to dictate how a given FOSS project prioritizes its work. You're not paying for anything, and so you're (still) getting what you paid for. Of course, you could rearchitect your own project to switch to another framework or language, but that sure sounds like a lot of work - definitely easier to just demand stuff...

> Speaking of, if you think you need python 3.13 today (or freethreading tomorrow, etc.)

That’s a make no sense when the majority of open source projects (especially the big ones) support python 3.13!

It’s not about supporting all its features. It’s only about some works to make it compatible.
Note that some open source projects already support python 3.14 while tensorflow need between 6 and 8 months to support it from it official release.

Why should I install specifically python 3.12 for tensorflow while all others support python 3.13?

Also I don’t think such big projects like tensorflow doesn’t get any supports or donations at all.

So all the arguments that you use make no sense.

**Comment by khteh** [comment]
right/wrong arguments don't advance the common interests of the project and it's users. There are 3 ways we could contribute to FOSS:

1) Use and test it. Create an issue here to bring it to the attention of active members of the project.
2) More capable? create a PR.
3) $$$

I have seen some even worse FOSS projects here in github - months or years since the last commit, no response or whatsoever to any issue / discussion created in the repos. Talking about "disregard", these maintainers definitely win the title. So, be grateful when the maintainers of the FOSS projects respond. Let's create a net-positive environment the FOSS projects to thrive - both as users and maintainers. 

And let's keep the tensor flowing!

**Comment by Aeonitis** [comment]
All, and I mean all contributors are doing a great job.

Delayed gratification is a learned art with Open Source, let's all remember that on both sides.

**Comment by muayyad-alsadi** [comment]
> More capable? create a PR.

The PR to make it work with py3.13 is already there (thank to archlinux users).

Again and again, what is preventing support for py3.13 is not technical. It's some politic

**Comment by muayyad-alsadi** [comment]
I think I mentioned this before. In the old days Ubuntu used to ship a year old kernel just because some lazy nvidia employee did not update closed source nvidia drivers to work with latest kernel.
We used to hate this, but we understand why it happens, because it's closer source. It's a church not a bazar.

This is not supposed to happen in opensource
It might happens once due to introduced breaking changes or major releases

But with TF it happens systematically every single minor release from PSF.

If I got a fresh device, mac, windows or linux (arch or fedora) having latest stable python from python foundation I should be able to install tf.

If once I could not do that due to some major breaking changes from PSF, we understand this.
But this should not be the normal process that happens every release.

There should be a way to unpin versions, get latest. If there are foreign structure there should be a script to update them.

The entire thread is about what is the work we need to do so we don't have to open such ticket evey 6 months with every new minor python release.

I'm not asking you to support every new feature in every new python release. Just rebuild the latest TF with latest Python from PSF. Don't be the old nvidia church. Do not hold us to a year old python.

**Comment by muayyad-alsadi** [comment]
> That is the same claim you made in #78774 (comment) to which I listed the ABI breaking changes that were breaking the build in #78774 (comment). The same ABI compatibility issues that were addressed in c0093bf.

Yes this commit/patch

https://github.com/tensorflow/tensorflow/commit/c0093bfb21f0e1157e1892cc8273e10831077174

Is basically do unpinning the requirements and syncing a foreign structure.

Let's have mutual respect, there are some people here who can read a patch and know what it does.

**Comment by swiftclouddbs** [comment]
Got a good chuckle about the entitlement, or sense of entitlement, some folks have about open source projects.  Had to wind back my project to use 3.12.  I have 3.10 still running other stuff in prod so not stressed.

Anyone want to help a new guy get up to speed? I am ready to post my project on GitHub.  Below is the tree.  It is software for picking trades in the options market.  Here is the tree:  
acropolis:/Users/dev/AlphaSeeker% tree -L 2
.
├── __pycache__
│   ├── config.cpython-312.pyc
│   └── config.cpython-313.pyc
├── alpha_seeker.db
├── config.py
├── daily_scripts
│   ├── __pycache__
│   ├── backfill_dte_on_options_data.py
│   ├── calculate_historical_volatility.py
│   ├── check_dte_on_options_data.py
│   ├── clean_tables.py
│   ├── daily_archive_expired_greeks.py
│   ├── daily_calculate_all_greeks_target_date.py
│   ├── daily_calculate_all_greeks_updated.py
│   ├── daily_calculate_all_greeks_useable.py
│   ├── daily_calculate_all_greeks.py
│   ├── daily_calculate_historical_volatility.py
│   ├── daily_data_quality_tag.py
│   ├── daily_historical.py
│   ├── daily_junk_contracts_tag.py
│   ├── daily_junk_removal.py
│   ├── daily_options_ingest.py
│   ├── daily_options_ingest.py.old
│   ├── daily_prescreen_contracts.py
│   ├── daily_remove_junk_contracts.py
│   ├── daily_rfr_pull.py
│   ├── daily_volatility_new.py
│   ├── daily_volatility.py
│   ├── eod_report_2.py
│   ├── eod_report_3.py
│   ├── eod_report.py
│   ├── get_last_market_date.py
│   ├── greeks_data.db
│   ├── old_scripts
│   ├── once_backfill_days_to_expiry.py
│   ├── rebuild_greeks.py
│   ├── run_daily_pipeline.sh
│   └── utils
├── dashboards
│   └── dashb_1.py
├── data
│   ├── =
│   ├── greeks_data.db
│   └── greeks.db
├── feature_store
│   ├── build_feature_store.py
│   ├── feature_store_update_orig.py
│   ├── feature_store_update.py
│   ├── features_definitions.sql
│   └── utils
├── main.py
├── portal_apps
│   ├── acropolis.sh
│   ├── gradio_bull_put_spread.py
│   ├── gradio_InfrastructureDashboard.py
│   ├── gradio_iV_hV_app.py
│   ├── gradio_mispricing_explorer.py
│   ├── gradio_option_margin_screen.py
│   ├── gradio_options_pie_chart.py
│   ├── gradio_ticker_count_pie.py
│   ├── gradio_trade_suggest_2.py
│   └── options.pid
├── README.md
├── reports
│   ├── credit_spread_candidates_con_ids.csv
│   ├── credit_spread_candidates_sample.csv
│   ├── credit_spread_report_2025-04-21.csv
│   ├── strategy_report.csv
│   ├── strategy_report.json
│   └── strategy_suggestions_balanced.csv
├── runner.py
├── trade_engine
│   ├── __init__.py
│   ├── __pycache__
│   ├── candidate_table_trade_simulator.py
│   ├── config.py
│   ├── feature_table_trade_simulator.py
│   ├── old_trade_simulator.py
│   ├── options_table_trade_simulator.py
│   ├── Trade_Engine_1.py
│   ├── Trade_Platform_1.py
│   ├── Trade_Sim_1.py
│   └── trade_simulator.py
├── trade_selection
│   ├── candidate_bs_diff.py
│   ├── candidate_filter_with_bs.py
│   ├── candidate_filter.py
│   ├── credit_spread_detector_old.py
│   ├── credit_spread_detector.py
│   ├── gradio_candidate_filter_with_bs.py
│   ├── gradio_candidate_llm_ranker_great.py
│   ├── gradio_candidate_llm_ranker_V2.py
│   ├── gradio_candidate_llm_ranker_V3.py
│   ├── gradio_candidate_llm_ranker.py
│   ├── gradio_candidate_viewer_good.py
│   ├── gradio_candidate_viewer.py
│   ├── gradio_LLM-SQL-QA.py
│   ├── gradio_LLM-trade_QA.py
│   ├── llm_candidate_trade_review.py
│   ├── llm_strategy_suggestions.py
│   ├── llm_trade_selector.py
│   ├── strategy_selector.py
│   ├── utils
│   ├── view_credit_spread_2.py
│   ├── view_credit_spread_3.py
│   └── view_credit_spreads.py
├── utils
│   └── logger.py
├── venv
│   ├── bin
│   ├── include
│   ├── lib
│   ├── pyvenv.cfg
│   └── share
└── venv312
    ├── bin
    ├── include
    ├── lib
    └── pyvenv.cfg

26 directories, 94 files
acropolis:/Users/dev/AlphaSeeker% 

Happy to make you (all) collaborators.  If you want to get familiar with the project even better!  Looking for some coaching to get it up and make simple modifications to the page.  I will do all the work, just looking for a few collaborators and tips to get the word out there is an open source project that will make a lot of BIG$CORPs happy.  

All the best!!

**Comment by mihaimaruseac** [comment]
It has always been the policy of TF to do a release that supports a new version of Python __after__ that release. This time the team slipped and missed one release, but @learning-to-play 's team already worked on this. Nightly already supports 3.13 and next release would support it too.

Please stop spamming with demands for support or we will be forced to lock the thread. There has been no PR submitted to the repository by people complaining here, the commit mentioned is from the OSS team.

**Comment by Maziar-fanaeipour** [comment]
This time the team slipped and missed one release

**Comment by swiftclouddbs** [comment]
Thank you all for your hard work.  Many, many thanks.
Pushing the limits with some tfdf libraries and went all the way back to Python 3.9.  Problem went away so I’ve stepped back up to Python 3.11 and … no problem.
Sometimes it IS programmer error.
You folks are the best!  Thanks!

Sent from Yahoo Mail for iPhone


On Sunday, May 11, 2025, 2:41 AM, Maziar-fanaeipour ***@***.***> wrote:

Maziar-fanaeipour left a comment (tensorflow/tensorflow#78774)
This time the team slipped and missed one release

**Comment by ralphrmartin** [comment]
It also seems that Debian triie will only have Python 3.13....

**Comment by vidakpop** [comment]
did you find a workaround apart from downgrading the python version to 3.12

**Comment by mihaimaruseac** [comment]
> did you find a workaround apart from downgrading the python version to 3.12

You can use `tensorflow-nightly` which has support for 3.13.

Or you can install alternative python versions on your system. Simplest approach is to use `hatch`, but there are other possibilities too.

**Comment by slittle95** [comment]
Do we have any idea when the stable build will include 3.13 support? I usually don't like to be this person, but I can't help noticing this issue has been open for 8 months. Just want to know whether I should commit my documentation updates with tf-nightly or just hold off.

**Comment by johnthagen** [comment]
For a cross-reference for the larger TF community of packages, `ai-edge-litert` (previously known as Tensorflow Lite) also does not yet support 3.13

- https://github.com/google-ai-edge/LiteRT/issues/469

**Comment by mihaimaruseac** [comment]
> Do we have any idea when the stable build will include 3.13 support? I usually don't like to be this person, but I can't help noticing this issue has been open for 8 months. Just want to know whether I should commit my documentation updates with tf-nightly or just hold off.

Next release, branch cut has not happened yet, but please watch the repository.

> For a cross-reference for the larger TF community of packages, `ai-edge-litert` (previously known as Tensorflow Lite) also does not yet support 3.13
> 
> * [Python 3.12 and 3.13 support google-ai-edge/LiteRT#469](https://github.com/google-ai-edge/LiteRT/issues/469)

This makes sense since the projects use to share similar release infra. But now they separate (are separating) so this should change

**Comment by mihaimaruseac** [comment]
TF 2.20.0-rc0 was just released and it [contains support for Python3.13](https://pypi.org/project/tensorflow/2.20.0rc0/#files). You can install and test it now with `python -m pip install tensorflow==2.20.0rc0` (or similar, dependending on what you use)

**Comment by khteh** [comment]
When is the GA release?

**Comment by mihaimaruseac** [comment]
On the past release I said:

> Next release of TF. Branch cut should occur by mid of February at most and final release at most a month after, assuming no delays.

In general, there was at most a month between RC0 and final release, even in the world when we were doing several RCs. After I left the team there's only RC0 and then final/GA (with very few exceptional RC1s) but even that took nearly a month. And, judging by the time between the branch cut and the RC0 release, it might be more this time?

**Comment by Stunkymonkey** [comment]
it just got released: https://pypi.org/project/tensorflow/2.20.0/

**Comment by mihaimaruseac** [comment]
Then we can finally close this issue. Thank you for patiently waiting.

**Comment by mweinelt** [comment]
https://www.tensorflow.org/install/pip#package_location still lists 2.19.0, but not 2.20.0. Can we expect an update there shortly?

**Comment by mihaimaruseac** [comment]
Please open a new issue to update the documentation. It seems there's a break in the process since several releases back and the team that does the releases has not yet resolved this issue
